framework/src/main/java/org/checkerframework/framework/util/DefaultQualifierKindHierarchy.java

```diff
@@ -0,0 +1,802 @@
+            }
+        }
+        for (DefaultQualifierKind qualifierKind : qualifierKinds) {
+            for (DefaultQualifierKind top : tops) {
+                if (qualifierKind.isSubtype(top)) {
+                    if (qualifierKind.top == null) {
+                        qualifierKind.top = top;
+                    } else if (qualifierKind.top != top) {
+                        throw new TypeSystemError(
+                                "Multiple tops found for qualifier %s. Tops: %s and %s.",
+                                qualifierKind, top, qualifierKind.top);
+                    }
+                }
+            }
+            if (qualifierKind.top == null) {
+                throw new TypeSystemError(
+                        "Qualifier %s isn't a subtype of any top. tops = %s", qualifierKind, tops);
+            }
+            qualifierKind.poly = qualifierKind.top.poly;
+        }
+        for (DefaultQualifierKind qualifierKind : qualifierKinds) {
+            for (DefaultQualifierKind bot : bottoms) {
+                if (bot.top != qualifierKind.top) {
+                    continue;
+                }
+                if (qualifierKind.bottom == null) {
+                    qualifierKind.bottom = bot;
+                } else if (qualifierKind.top != bot) {
+                    throw new TypeSystemError(
+                            "Multiple bottoms found for qualifier %s. Bottoms: %s and %s.",
+                            qualifierKind, bot, qualifierKind.bottom);
+                }
+                if (qualifierKind.isPoly()) {
+                    assert bot.strictSuperTypes != null
+                            : "@AssumeAssertion(nullness):strictSuperTypes should be nonnull.";
+                    bot.strictSuperTypes.add(qualifierKind);
+                }
+            }
+            if (qualifierKind.bottom == null) {
+                throw new TypeSystemError(
+                        "Cannot find a bottom qualifier for %s. bottoms = %s",
+                        qualifierKind, bottoms);
+            }
+        }
+    }
+
+    /**
+     * Returns the set of all qualifier kinds that are a strict supertype of {@code qualifierKind}.
+     *
+     * @param qualifierKind the qualifier kind whose super types should be returned
+     * @param directSuperMap directSuperMap
+     * @return the set of all qualifier kinds that are a strict supertype of {@code qualifierKind}
+     */
+    private Set<QualifierKind> findAllTheSupers(
+            @UnderInitialization DefaultQualifierKindHierarchy this,
+            @KeyFor("#2") QualifierKind qualifierKind,
+            Map<DefaultQualifierKind, Set<DefaultQualifierKind>> directSuperMap) {
+        Queue<DefaultQualifierKind> queue = new ArrayDeque<>(directSuperMap.get(qualifierKind));
+        Set<QualifierKind> allSupers = new TreeSet<>(directSuperMap.get(qualifierKind));
+        Set<DefaultQualifierKind> visited = new HashSet<>();
```

Comment:
I don't think so. A `QualifierKind` is added to `visited` once its super qualifier kinds have been added to `allSupers`.  T `allSupers` contains `QualifierKind`s whose super qualifier kinds have not yet been added to allSupers.   But maybe I'm misunderstanding your comment.

Reply:
The code as is works, but I was wondering if it could be simplified.  Maybe a comment about the relationship between the variables is enough to help people understand the code.